src/Readers/Vipr.Reader.OData.v4/OdcmReader.cs

```diff
@@ -738,6 +742,22 @@ private void WriteProperty(OdcmClass odcmClass, IEdmProperty property)
                     AddVocabularyAnnotations(odcmProperty, property);
 
                     odcmClass.Properties.Add(odcmProperty);
+
+
+                    var derivedTypes = _edmModel.GetDerivedTypeConstraints(property);
+                    if (derivedTypes != null)
+                        foreach (var derivedType in derivedTypes)
+                        {
+                            var odcmDerivedClass = TryResolveType<OdcmClass>(derivedType);
+
+                            var derivedCastProperty = odcmProperty.Clone($"{odcmProperty.Name}As{odcmDerivedClass.Name.First().ToString().ToUpper()}{odcmDerivedClass.Name.Substring(1)}");

```

Comment:
I think this affects the decision as to if we are going to use generic type casts in other languages or if we are going to follow the same design from Java.

Since we haven't designed this feature for other languages yet, I would suggest keeping the name of these properties in a data structure in Odcm model and generating `propertyAsDerived` naming in the specific Java logic in the SDK generator.

cc: @MIchaelMainer

Reply:
I've spent some time trying to get this code section out of viper into typewriter but doing this breaks a lot of encapsulation (access to all the IEdm stuff, finding some IEdm types and annotations, accessing the property cache...).
So I just pushed my counter proposition: leave it in vipr but enable it based on a flag. And the flag is set by typewriter based on the language. (only java currently)
What do you think of that alternative?

Reply:
That is a good alternative.